### PR TITLE
Lots more Holidays, Future Proofs Easter Calculations

### PR DESCRIPTION
**Pull request by RemieRichards**
- Easter Calculations will now work until 4099, instead of breaking in 2040. ( @MrPerson was lazy (iirc it was him)), As a quirk of the algorithm, it can also calculate Easter Dates all the way back to 1583, I doubt we'll use that though
- Adds Space Easter, Space Easter is set on the date that easter would fall in SS13's universe, this is calculated using SS13's in-game year (CURRENT YEAR + 540), this use the normal easter events (bunnies etc.) and will probably overlap some-what with CURRENT YEAR easter, but it's just bunnies and stuff so probably not a concern
- Adds Mothering Sunday, always 3 weeks (21 days) before Easter (hence improving easter calculations)
- Adds Ash Wednesday
- Adds Good Friday
- Adds Pancake Day / Mardi Gras / Shrove Tuesday

See @Really-Good-Soda-Flavor, it's not hard!
Will conflict with #28054 but I'll fix conflicts when/if they happen.

### CODE REVIEW

File: code/modules/holiday/holidays.dm [hunk 1]
```diff
@@ -420,11 +384,37 @@
...
+	if(!begin_month)
+		var/year = text2num(time2text(world.timeofday, "YYYY"))
+		var/list/motheringSundayResults = MotheringSundayDate(year)
+		end_day = begin_day = motheringSundayResults["day"]
+		end_month = begin_month = motheringSundayResults["month"]
+
+	return ..()
+
+/datum/holiday/mothering_sunday/greet()
+	return "It's Mothering Sunday! the -REAL- Mother's Day, you did remember to get her a gift right?"
```
Comment (Xhuis) [inline code/modules/holiday/holidays.dm]: let's keep stirring up trouble between people living in different countries by claiming that one holiday is inferior to another variety instead!

Comment (DoctoreMad) [inline code/modules/holiday/holidays.dm]: >Oh no the players might metagrudge each other due to horrible ethnocentrism

Comment (Xhuis) [inline code/modules/holiday/holidays.dm]: I'm not worried about the players, I just think that it's a little immature to be claiming superiority over something as minor as a holiday.

Comment (DoctoreMad) [inline code/modules/holiday/holidays.dm]: Is this what people say when they realize their holiday is OBJECTIVELY INFERIOR?

smugcat

Comment (Xhuis) [inline code/modules/holiday/holidays.dm]: Oh, right, I forgot rational discussion wasn't possible with children. My bad.

**Comment by bgobandit** [comment]
my space mother died fuck you

**Comment by RemieRichards** [comment]
@bgobandit legitimately, that was one of the reasons I contemplated not adding ```greet()``` text, and just having it fall to the default "Have a nice [HOLIDAY]"

**Comment by Really-Good-Soda-Flavor** [comment]
Wow, all of that pedantic complaining about how it's technically Mother's Day and then you still call it Mothering Sunday.

In all seriousness though, well done, I imagine it was not easy to do for whoever wrote the BASIC algorithm, and **now you should add Ash Wednesday and Mardi Gras as well** @RemieRichards

File: code/modules/holiday/holidays.dm (hunk 1 above)
Comment (Really-Good-Soda-Flavor) [inline code/modules/holiday/holidays.dm]: You didn't capitalize the new sentence, you put dashes around REAL when they shouldn't be there. Come on, man.

**Comment by RemieRichards** [comment]
@Really-Good-Soda-Flavor that's kind of the joke.
Are those two related to easter? (as in, simple offsets like how Mothering Sunday is Easter-21 days), or are they just similar "these are complex cos calenders suck!"

**Comment by Really-Good-Soda-Flavor** [comment]
All of them are based off of the month of Lent, and there are a bunch of other holidays that are too, like Good Friday

File: code/modules/holiday/holidays.dm [hunk 2]
```diff
@@ -420,11 +384,37 @@
...
 	return ..()
 
 /datum/holiday/easter/celebrate()
 	..()
 	GLOB.maintenance_loot += list(
 		/obj/item/weapon/reagent_containers/food/snacks/egg/loaded = 15,
 		/obj/item/weapon/storage/bag/easterbasket = 15)
+
+/datum/holiday/easter/greet()
+	return "Greetings! in the far flung past of [current_year], today would be easter, Be sure to have a happy one and keep an eye out for Bunnies!"
```
Comment (Really-Good-Soda-Flavor) [inline code/modules/holiday/holidays.dm]: Too meta

**Comment by Really-Good-Soda-Flavor** [comment]
Actually, Lent is based on Easter. It starts on Ash Wednesday, 46 days before Easter Sunday @RemieRichards 

Good Friday is 2 days before Easter

Mardi Gras is 47 days before Easter, though it is usually celebrated for an entire week in advance

**Comment by RemieRichards** [comment]
I will certainly look into those then.

**Comment by Xhuis** [comment]
If you mow the lawn on mardi gras is it mardi grass?

**Comment by Jazaen** [comment]
Goes back to only 1583? I wanted to host spessmens at marriage ceremony of Shakespeare and Anne Hathaway. Literally unplayable.

**Comment by RemieRichards** [comment]
Got AshWednesday almost done, but got an off by 1 error in it, so that will suck to fix.

**Comment by MrPerson** [comment]
LAZY????? I had to look up those Easter dates and input them MANUALLY

**Comment by Grazyn** [comment]
Still no Ramadan

**Comment by Really-Good-Soda-Flavor** [comment]
I'm working on Ramadan

**Comment by RemieRichards** [comment]
LEAP YEARS.
OF COURSE, WHY'D IT HAVE TO BE LEAP YEARS.

**Comment by RemieRichards** [comment]
K, leap years done, Ash Wednesday and Shrove Tuesday (Mardi Gras) are done.
Going to do good friday too since it's just easter-2

**Comment by nicbn** [comment]
YAY IS THIS THE SPACE RACISM PR

**Comment by Grazyn** [comment]
Throwing a couple more easy-to-add religious holidays
Pentecost (big thing in Germany): 50 days after Easter
Immaculate conception: December 8

**Comment by RemieRichards** [comment]
@Grazyn

Pentecost is a no no, 50 days after easter is the same as the 46, and 47 of Ash wednesday and Shrove tuesday, as in: annoying to calculate.
Immaculate conception is just a fixed date.

I might pick them up some other time, but for now, this PR is finalised.

**Comment by nicbn** [comment]
We could also add [Orthodox Valentine's Day](https://en.wikipedia.org/wiki/Valentine%27s_Day) (July 6), [Father's Day](https://en.wikipedia.org/wiki/Father%27s_Day) (March 19, but also March 18 in many countries), [Talk Like a Pirate Day](https://en.wikipedia.org/wiki/International_Talk_Like_a_Pirate_Day) (September 19), [Monkey Day](https://en.wikipedia.org/wiki/Monkey_Day) (December 14), [Cosmonautics Day](https://en.wikipedia.org/wiki/Cosmonautics_Day) (April 12), [St. Patrick's Day](https://en.wikipedia.org/wiki/Saint_Patrick%27s_Day) (March 17) etc.
Also please add the holiday names at misc.dm, that way we can know what we've got

**Comment by ShizCalev** [comment]
>monkey day

what

**Comment by Grazyn** [comment]
Definitely need Monkey Day and St. Paddy's at the very least

File: code/modules/holiday/holidays.dm [hunk 3]
```diff
@@ -420,11 +384,80 @@
...
```
Comment (Cyberboss) [inline code/modules/holiday/holidays.dm]: Capitalization

File: code/modules/holiday/holidays.dm [hunk 4]
```diff
@@ -420,11 +384,80 @@
...
+
+
+/datum/holiday/easter/spess
+	name = SPACE_EASTER
+	year_offset = 540 //Canonicially the year is CURRENT YEAR + 540, so this is Easter as it would be IN SS13 itself
+
+/datum/holiday/easter/spess/greet()
+	return "Greetings! the year is [current_year+year_offset], and it's easter! be sure to have a happy one and keep an eye out for Easter Bunnies!"
```
Comment (Cyberboss) [inline code/modules/holiday/holidays.dm]: Capitalization

File: code/modules/holiday/holidays.dm [hunk 5]
```diff
@@ -420,11 +384,80 @@
...
+	if(!begin_month)
+		var/year = text2num(time2text(world.timeofday, "YYYY"))
+		var/list/motheringSundayResults = MotheringSundayDate(year)
+		end_day = begin_day = motheringSundayResults["day"]
+		end_month = begin_month = motheringSundayResults["month"]
+
+	return ..()
+
+/datum/holiday/mothering_sunday/greet()
+	return "It's Mothering Sunday! the -REAL- Mother's Day, you did remember to get her a gift right?"
```
Comment (Cyberboss) [inline code/modules/holiday/holidays.dm]: Capitalization

File: code/modules/holiday/holidays.dm [hunk 6]
```diff
@@ -420,11 +384,80 @@
...
+		var/year = text2num(time2text(world.timeofday, "YYYY"))
+		var/list/shroveTuesdayResults = ShroveTuesdayDate(year)
+		end_day = begin_day = shroveTuesdayResults["day"]
+		end_month = begin_month = shroveTuesdayResults["month"]
+
+	return ..()
+
+
+/datum/holiday/pancakeday/greet()
+	return "Today is Pancake day! also known as Mardi Gras and Shrove Tuesday, but who cares about that, PANCAAAAAKES!!!1!!"
```
Comment (Cyberboss) [inline code/modules/holiday/holidays.dm]: Capitalization

**Review by Cyberboss — DISMISSED** [review]

File: code/modules/holiday/holidays.dm (hunk 3 above)
Comment (RemieRichards) [inline code/modules/holiday/holidays.dm]: Easter? Bunnies?

File: code/modules/holiday/holidays.dm (hunk 4 above)
Comment (RemieRichards) [inline code/modules/holiday/holidays.dm]: Easter? Bunnies?

Comment (tortellinitony) [inline code/modules/holiday/holidays.dm]: Easter needs to always be capitalized and bunnies does not have to be capitalized

Comment (RemieRichards) [inline code/modules/holiday/holidays.dm]: Capitalised bunnies comes form the old message.

File: code/modules/holiday/holidays.dm (hunk 3 above)
Comment (nicbn) [inline code/modules/holiday/holidays.dm]: Be -> be
in -> In
easter -> Easter
Bunnies -> bunnies

File: code/modules/holiday/holidays.dm (hunk 4 above)
Comment (nicbn) [inline code/modules/holiday/holidays.dm]: the -> The
be -> Be
Easter Bunnies is ok

File: code/modules/holiday/holidays.dm (hunk 5 above)
Comment (nicbn) [inline code/modules/holiday/holidays.dm]: the -> The
But please remove the whole "the -REAL- Mother's Day, " sentence and capitalize "you".

File: code/modules/holiday/holidays.dm (hunk 6 above)
Comment (nicbn) [inline code/modules/holiday/holidays.dm]: Pancake day -> Pancake Day OR pancake day
also -> Also

File: code/__DEFINES/misc.dm [hunk 7]
```diff
@@ -38,6 +38,7 @@ Will print: "/mob/living/carbon/human/death" (you can optionally embed it in a s
 #define VALENTINES				"Valentine's Day"
```
Comment (nicbn) [inline code/__DEFINES/misc.dm]: Add all the holidays here, for the sake of reviewers and future coders.

**Review by nicbn — CHANGES_REQUESTED** [review]

File: code/__DEFINES/misc.dm (hunk 7 above)
Comment (RemieRichards) [inline code/__DEFINES/misc.dm]: No, that's not what this is for, it's for checking against holiday strings, not for just listing all holidays.
Technically I shouldn't have added the ```SPACE_EASTER``` one but I figured someone might want to make Space Easter slightly different to Easter at some point, so I pre-emptively added it.

Comment (nicbn) [inline code/__DEFINES/misc.dm]: Sure, but defining it all here is a really better way to do it than hardcoding all the holidays names.
Also, as define is preprocessed, there should be no performance issues.

In other words, adding the holidays here is a good practice (the best practice is to create a JSON containing holiday names, descriptions and dates).

**Comment by lzimann** [comment]
@RemieRichards conflicts

**Comment by Cyberboss** [comment]
@RemieRichards !

**Comment by RemieRichards** [comment]
@Cyberboss already explained myself in the OP, doesn't matter they're unrelated conflicts, I'm not going to fix conflicts when I'm near-guaranteed to get another set very soon after.

I will fix them both, in time.

**Comment by lzimann** [comment]
@RemieRichards reeeeeeeeeeeeeemie

**Comment by KorPhaeron** [comment]
@RemieRichards 

Let me know when you are done

**Pull request closed without merging**